File: xconf-angular-admin/src/main/webapp/app/shared/changelog/changelog.controller.js
/*******************************************************************************
 * If not stated otherwise in this file or this component's Licenses.txt file the
 * following copyright and licenses apply:
 *
 * Copyright 2018 RDK Management
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
(function() {
    'use strict';

    angular
        .module('app.changeLog')
        .controller('ChangeLogController', controller);

    controller.$inject = ['$log', 'changeLogService'];

    function controller($log, changeLogService) {
        var vm = this;
        vm.intervalsList = [];

        init();

        function init() {
            changeLogService.getChangeLog().then(
                function(result) {
                    fillIntervalsList(result.data);
                },
                function(reason) {
                    $log.debug('REASON', reason);
                }
            );
        }

        function fillIntervalsList(data) {
            //each timestamp corresponds to list of changes
            var sortedTimestamps = getSortedKeys(data);
            var startOfInterval;
            var endOfInterval;
            for(var i = 0; i < sortedTimestamps.length; i++) {
                var changesObj = {};
                startOfInterval = sortedTimestamps[i];
                if (i === sortedTimestamps.length - 1) {
                    endOfInterval = new Date();
                } else {
                    //end of current interval one second less than start of next interval
                    endOfInterval = sortedTimestamps[i+1] - 1000;
                }
                changesObj.startOfInterval = timeFromTimestamp(startOfInterval);
                changesObj.endOfInterval = timeFromTimestamp(endOfInterval);
                changesObj.changesList = data[startOfInterval];
                changeLogService.countAndSaveOperationsNumber(data[startOfInterval], changesObj);
                vm.intervalsList.push(changesObj);
            }
        }

        function getSortedKeys(object) {
            var sortedKeys = [];
            for (var k in object) {
                if (object.hasOwnProperty(k))
                {
                    sortedKeys.push(parseInt(k));
                }
            }
            sortedKeys.sort(function(a, b) {
                return a - b;
            });

            return sortedKeys;
        }

        function timeFromTimestamp(timestamp) {
            var date = new Date(timestamp);
            var regexp = /\((.)*\)/;
            var result = date.toTimeString().replace(regexp, '');
            return result;
        }
    }
})();

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/controller/shared/ChangeLogController.java
/*******************************************************************************
 * If not stated otherwise in this file or this component's Licenses.txt file the
 * following copyright and licenses apply:
 *
 * Copyright 2018 RDK Management
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package com.comcast.xconf.admin.controller.shared;

import com.comcast.xconf.admin.service.shared.ChangeLogService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping(ChangeLogController.URL_MAPPING)
public class ChangeLogController {

    public static final String URL_MAPPING = "api/changelog";

    private ChangeLogService changeLogService;

    @Autowired
    public ChangeLogController(ChangeLogService changeLogService) {
        this.changeLogService = changeLogService;
    }

    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
    public Map<Long, List<Change>> getChangeLog() {
        return changeLogService.getChangeLog();
    }

    public static class Change {
        String changedKey;
        String operationType;
        String cfName;
        private String userName;

        public Change() {}

        public Change(String changedKey, String operationType, String cfName, String userName) {
            this.changedKey = changedKey;
            this.operationType = operationType;
            this.cfName = cfName;
            this.userName = userName;
        }

        public String getChangedKey() {
            return changedKey;
        }

        public void setChangedKey(String changedKey) {
            this.changedKey = changedKey;
        }

        public String getOperationType() {
            return operationType;
        }

        public void setOperationType(String operationType) {
            this.operationType = operationType;
        }

        public String getCfName() {
            return cfName;
        }

        public void setCfName(String cfName) {
            this.cfName = cfName;
        }

        public String getUserName() {
            return userName;
        }

        public void setUserName(String userName) {
            this.userName = userName;
        }
    }
}